share/python/procedures/findif_response_utils/tensor_helper.py

```diff
@@ -0,0 +1,118 @@
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# @END LICENSE
+#
+
+"""
+Module of helper functions for distributed ccresponse computations.
+
+Defines functions for retrieving tensors computed at displaced geometries.
+"""
+from __future__ import absolute_import
+from __future__ import print_function
+import collections
+import shelve
+import copy
+import os
+import psi4
+import p4util
+from p4const import *
+
+
+def synthesize_displaced_tensor(db, signature, row_dim):
+    """
+        Gathers a list of tensors, one at each displaced geometry.
+
+    db: (database) the database object for this property calculation
+    signature: (string) The string that notifies the matrix reader that the
+        targeted tensor data begins.
+    row_dim: the expected number of rows that this value should be printed
+        across in the file
+
+    Returns a 2d list result[i][j]:
+        i: indexes displacements
+        j: indexes elements of the flattened tensor at some displacement
+    Throws: none
+    """
+    result = []
+    for job in db['job_status']:
+        with open('{}/output.dat'.format(job)) as outfile:
+            result.append(grab_psi4_matrix(outfile, signature, row_dim))
+
+    return result
+
+    # END sythesize_displaced_tensor()
+
+
+def grab_psi4_matrix(outfile, matrix_name, row_tot):
+    """
+        Grabs a psi4 matrix printed to the output file
+
+    outfile: ( file ) handle open in read mode, where the data should be found
+    matrix_name: ( string ) that indicates the matrix data is found on the lines
+        below
+    row_tot: ( int ) indicates the number of lines that the matrix data should
+        be printed across in the file
+
+    Returns: matrix_data a list of matrix elements, len = 3*row_tot
+
+    Throws: Exception (Collecting matrix data failed) if
+            It is found, but not the correct size,
+            It found the matrix header, but no data,
+            It can't find the matrix header in the file
+    """
+    collect_matrix = False
+    n_rows = 0
+    n_tries = 0
+    matrix_data = []
+    for line in outfile:
+        if matrix_name in line:
+            collect_matrix = True
+        if collect_matrix and (n_rows < row_tot):
+            try:
+                n_tries += 1
+                if n_tries > (row_tot + 13):
+                    raise Exception('{} Matrix was unreadable. Scanned {}'
+                                    'lines.'.format(matrix_name, n_tries))
+                else:
+                    (index, x, y, z) = line.split()
+                    matrix_data.append(float(x))
+                    matrix_data.append(float(y))
+                    matrix_data.append(float(z))
+                    n_rows += 1
+            except:
+                pass
+        if (n_rows == row_tot) and (len(matrix_data) != 3 * row_tot):
+            raise Exception('Collecting {} matrix data failed!'
```

Comment:
You're right that `ValidationError` sounds input-specific, but I've used it pretty generically, and you're welcome to follow suit. If you'd like to make a more directed error, just follow the printing convention [here](https://github.com/psi4/psi4/blob/master/share/python/p4util/exceptions.py#L71-L89) so that messages go to screen and outfile. My view at present is just to make any error classes you need, especially if you want to catch them, otherwise freely default to `ValidationError`. Then in future when we've a better handle on use cases, can do any cleanup and consolidation.